src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java

```diff
@@ -375,38 +376,37 @@ public void registerToolchains(Sequence<?> toolchainLabels) throws EvalException
       },
       useStarlarkThread = true)
   public ModuleExtensionProxy useExtension(
-      String extensionBzlFile, String extensionName, boolean devDependency, StarlarkThread thread)
-      throws EvalException {
-    ModuleExtensionProxy newProxy =
-        new ModuleExtensionProxy(extensionBzlFile, extensionName, thread.getCallerLocation());
+      String extensionBzlFile, String extensionName, boolean devDependency, StarlarkThread thread) {
+    ModuleExtensionProxyBuilder newProxyBuilder = new ModuleExtensionProxyBuilder(extensionBzlFile,
+        extensionName, thread.getCallerLocation());
 
     if (ignoreDevDeps && devDependency) {
       // This is a no-op proxy.
-      return newProxy;
+      return newProxyBuilder.getProxy(devDependency);
     }
 
-    // Find an existing proxy object corresponding to this extension.
-    for (ModuleExtensionProxy proxy : extensionProxies) {
-      if (proxy.extensionBzlFile.equals(extensionBzlFile)
-          && proxy.extensionName.equals(extensionName)) {
-        return proxy;
+    // Find an existing proxy builder corresponding to this extension.
+    for (ModuleExtensionProxyBuilder proxyBuilder : extensionProxyBuilders) {
+      if (proxyBuilder.extensionBzlFile.equals(extensionBzlFile)
+          && proxyBuilder.extensionName.equals(extensionName)) {
+        // All proxies returned by withDevDependency share a common set of imports and tags.
+        return proxyBuilder.getProxy(devDependency);
       }
     }
 
     // If no such proxy exists, we can just use a new one.
-    extensionProxies.add(newProxy);
-    return newProxy;
+    extensionProxyBuilders.add(newProxyBuilder);
+    return newProxyBuilder.getProxy(devDependency);
   }
 
-  @StarlarkBuiltin(name = "module_extension_proxy", documented = false)
-  class ModuleExtensionProxy implements Structure {
+  class ModuleExtensionProxyBuilder {
```

Comment:
nit: I think this should be named "ModuleExtensionUsageBuilder". It's not exactly building proxies, but `ModuleExtensionUsage` objects (in fact it's really just a wrapper around a `ModuleExtensionUsage.Builder`; see `#buildUsage`). It just offers a proxy for Starlark code to use.

Reply:
Makes sense, done.